File: cli/provider/discovery.go
package provider

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"syscall"
)

type TransportDesc struct {
	Type string `json:"type"`
	Path string `json:"path,omitempty"`
	URL  string `json:"url,omitempty"`
}

type Descriptor struct {
	ID           string        `json:"id"`
	Name         string        `json:"name"`
	Version      string        `json:"version,omitempty"`
	SlopVersion  string        `json:"slop_version"`
	Transport    TransportDesc `json:"transport"`
	PID          int           `json:"pid,omitempty"`
	Capabilities []string      `json:"capabilities"`
	Description  string        `json:"description,omitempty"`
}

// Discover scans provider descriptor directories and returns live providers.
func Discover() ([]Descriptor, error) {
	var dirs []string

	// ~/.slop/providers/
	if home, err := os.UserHomeDir(); err == nil {
		dirs = append(dirs, filepath.Join(home, ".slop", "providers"))
	}

	// /tmp/slop/providers/
	dirs = append(dirs, filepath.Join(os.TempDir(), "slop", "providers"))

	var providers []Descriptor
	seen := map[string]bool{}

	for _, dir := range dirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}

		for _, entry := range entries {
			if entry.IsDir() || filepath.Ext(entry.Name()) != ".json" {
				continue
			}

			data, err := os.ReadFile(filepath.Join(dir, entry.Name()))
			if err != nil {
				continue
			}

			var desc Descriptor
			if json.Unmarshal(data, &desc) != nil {
				continue
			}

			if desc.ID == "" || seen[desc.ID] {
				continue
			}

			// Validate PID liveness
			if desc.PID > 0 && !isProcessAlive(desc.PID) {
				continue
			}

			seen[desc.ID] = true
			providers = append(providers, desc)
		}
	}

	sort.Slice(providers, func(i, j int) bool {
		return providers[i].Name < providers[j].Name
	})

	return providers, nil
}

// Address returns a display-friendly address for the descriptor.
func (d Descriptor) Address() string {
	switch d.Transport.Type {
	case "unix":
		return "unix:" + d.Transport.Path
	case "ws":
		return d.Transport.URL
	default:
		return d.Transport.Type
	}
}

func isProcessAlive(pid int) bool {
	process, err := os.FindProcess(pid)
	if err != nil {
		return false
	}
	err = process.Signal(syscall.Signal(0))
	return err == nil
}

File: apps/desktop/src-tauri/src/provider/discovery.rs
use std::path::PathBuf;

use serde_json::Value;

/// Scan discovery directories for SLOP provider descriptor files.
pub fn scan_providers() -> Vec<Value> {
    let mut providers = Vec::new();

    let dirs = discovery_dirs();
    for dir in dirs {
        if !dir.exists() {
            continue;
        }
        if let Ok(entries) = std::fs::read_dir(&dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().map_or(false, |ext| ext == "json") {
                    if let Ok(content) = std::fs::read_to_string(&path) {
                        if let Ok(val) = serde_json::from_str::<Value>(&content) {
                            providers.push(val);
                        }
                    }
                }
            }
        }
    }

    providers
}

/// Return the directories to watch for provider descriptors.
pub fn discovery_dirs() -> Vec<PathBuf> {
    let mut dirs = Vec::new();

    if let Some(home) = dirs::home_dir() {
        dirs.push(home.join(".slop").join("providers"));
    }

    dirs.push(PathBuf::from("/tmp/slop/providers"));

    dirs
}
